### DISCUSSION TITLE
What's the reason for moving "common" script in scriptblocks to next one if next one is considered as "baseScript"?

### DISCUSSION THREAD

**Discussion by Lkopo** [mpdf/mpdf#1816] [Q&A]
Hello,

I am fighting with issue when asian & non-asian characters are combined together. Example:

```
ンターJA健康ー康
```

with `autoScriptToLang` & `autoLangToFont` set to `true`, I get the following result:

![image](https://user-images.githubusercontent.com/6184484/217010487-aa41f5e0-9fb9-4a42-b90f-68141622ae49.png)

You can see the first `ー` character is not displayed properly. When I auto-force language on that text:

```html
<span lang="ja">
    ンターJA健康
</span>
```

This is what I get

![image](https://user-images.githubusercontent.com/6184484/217010149-41a5c2cf-f2cb-4783-9aa8-a606506dfc63.png)

Now it's printed out, but you can see the first "dash" is more bold than the second one. Guessing different font is being used for it.

To put this right - I don't have control of the input, I can have a control over template, but input is coming from other source. So I cannot put HTML tags into it based on the text value.

Coming to the question, by digging into the code, I found the following lines in Mpdf class https://github.com/mpdf/mpdf/blob/development/src/Mpdf.php#L26860-L26869:

```php
				// If scriptblock[x] = common & non-baseScript
				// and scriptblock[x+1] = baseScript
				// Move common script from end of x to start of x+1
				for ($sch = 0; $sch < $subchunk; $sch++) {
					if ($scriptblocks[$sch] > 0 && $scriptblocks[$sch] != $this->baseScript && $scriptblocks[$sch + 1] == $this->baseScript) {
						$end = count($chardata[$sch]) - 1;
						while ($chardata[$sch][$end]['script'] == 0 && $end > 1) { // common script
							$tmp = array_pop($chardata[$sch]);
							array_unshift($chardata[$sch + 1], $tmp);
							$end--;
						}
					}
				}
```

Where by looking and debugging it, I found out that first `$chardata` blocks were split correctly into groups of:

- `ンター` - evaluated as non-baseScript, lang is determined ✔
- `JA`  - evaluated as baseScript, no lang is set ✔
- `健康` - evaluated as non-baseScript, lang is determined ✔

But after evaluation of the following part, groups are changed:

- `ンタ` - evaluated as non-baseScript, lang is determined ✔
- `ーJA` - evaluated as baseScript, no lang is set ❌
- `健康` - evaluated as non-baseScript, lang is determined ✔

And I don't get it - why? This is eventually the cause for the issues I have when last non-latin character before latin is misinterpreted as basic latin character and therefore it's not visible with basic fonts. When I remove this part of the code locally, I am able to get the desired result:

![image](https://user-images.githubusercontent.com/6184484/217012190-990298a0-410b-4278-be4c-27d387969d7f.png)

I also checked other parts of the PDF and everything seems to be rendered fine.

I guess it was not put there randomly, there is some reason behind it. Therefore I am not confident to make a pull request with removed those lines without further knowledge of their real purpose. Can you please explain?

**Comment by finwe** [comment]
I am afraid only the original author of the library (or of these lines of code) could explain. I, as a lowly maintainer, simply do not know.

**Reply by Lkopo** [reply]
Thanks for the reply! Is the original author no longer reachable? Or what next steps would you suggest in order to fix the issue in correct manner? Thank you!

**Reply by finwe** [reply]
I am afraid the original author is disconnected from the project since it was (with his permission) taken over by the community almost 8 years ago. I cannot recommend a course of action as I have no idea how managing CJK scripts works, sorry. If you proposed a PR with an extended reasoning (you seem to understand this bit of code much more than I do), I could try to review it to the best of my abilities.

**Reply by Lkopo** [reply]
Okay so with deeper analysis, I think I understand the meaning of it. There are unicode characters which belongs to "common" script category, like currency signs (https://www.compart.com/en/unicode/U+20BF#UNC_SCRIPTS), emojies, etc. That "unlucky" character is also considered as common https://www.compart.com/en/unicode/U+30FC as well as in current UCDN database in project. (which is eventually out-dated, latest released version is 12.0.0, project uses 6.2.0, but upgrading it won't fix it either as categorization is correct). I assume that intentions were to leave common characters to "default" language so it looks more natural in default font. But some "common" characters are not contained in the fonts used for latin characters (not sure whether they need to be updated or are up-to-date enough).

So removing those suggested lines might lead to the unwanted visual changes for other documents where mixed with basic latin and other characters.

I was wondering about other options which leads to another question - is it somehow possible that when I specify language for some HTML element, autoScriptToLang will be disabled for that element? It breaks the styling when for some unknown reasons, it ignores `font-weight: bold` style even with `!important` set to it.

**Reply by Lkopo** [reply]
But also the author's decision isn't like a solution for every problem as this discussion proves it. Probably there is no correct approach how to handle such situation. Some case might require to use font based on previous characters and maybe for some cases it's required to use font for upcoming character. Probably the most safe approach here is to force only one language without `autoScriptToLang` for that element.

/e I've created a feature request https://github.com/mpdf/mpdf/issues/1817. I am working on a PR with some simple solution to achieve that.
